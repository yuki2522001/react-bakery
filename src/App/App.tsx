import React, { lazy, Suspense } from "react";
import { memo } from "react";
import { SWRConfig } from "swr";
import { Toaster } from "react-hot-toast";
import { Route, Routes } from "react-router-dom";
import { DataProvider } from "@context/DataContext";
import Categories from "@components/Categories/Categories";
import Footer from "@components/common/Footer/Footer";
import Header from "@components/common/Header/Header";
import Navigation from "@components/common/Navigation/Navigation";
import Loading from "@components/common/Loading/Loading";

// Lazy-loaded
const ProductListView = lazy(
  () => import("@pages/ProductListView/ProductListView")
);
const ProductGridView = lazy(
  () => import("@pages/ProductGridView/ProductGridView")
);
const ProductDetails = lazy(() => import("@pages/ProductDetail/ProductDetail"));

const App: React.FC = memo(() => {
  return (
    // The context SWRConfig can provide global configurations (options) for all SWR hooks.
    <SWRConfig value={{ revalidateOnFocus: false }}>
      <Header />
      <DataProvider>
        <Navigation />
        <Categories />
        {/* Suspense: Show a <Loading /> while the product list is loading */}
        <Suspense fallback={<Loading />}>
          <Routes>
            <Route path="/" element={<ProductListView />} />
            <Route path="/products" element={<ProductGridView />} />
            <Route path={`/products/:id`} element={<ProductDetails />} />
          </Routes>
        </Suspense>
      </DataProvider>
      <Footer />
      {/* Toaster: library show message */}
      <Toaster position="top-center" />
    </SWRConfig>
  );
});

export default App;
