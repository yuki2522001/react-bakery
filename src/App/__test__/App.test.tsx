import { render, screen } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import "@testing-library/jest-dom";
import App from "../App";

describe("App component", () => {
  test("rendering a view product list component that uses useLocation", () => {
    const history = createMemoryHistory();
    const route = "/";
    history.push(route);
    render(
      <Router location={history.location} navigator={history}>
        <App />
      </Router>
    );
    expect(screen.getByTestId("loading-page")).toBeInTheDocument();
  });

  test("rendering a product list component that uses useLocation", () => {
    const history = createMemoryHistory();
    const route = "/products";
    history.push(route);
    render(
      <Router location={history.location} navigator={history}>
        <App />
      </Router>
    );
    expect(screen.getByTestId("loading-page")).toBeInTheDocument();
  });

  test("rendering a product detail component that uses useLocation", () => {
    const history = createMemoryHistory();
    const route = "/products/:id";
    history.push(route);
    render(
      <Router location={history.location} navigator={history}>
        <App />
      </Router>
    );
    expect(screen.getByTestId("loading-page")).toBeInTheDocument();
  });

  test("matches snapshot", () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <Router location={history.location} navigator={history}>
        <App />
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
