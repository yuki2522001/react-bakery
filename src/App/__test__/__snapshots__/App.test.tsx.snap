// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component matches snapshot 1`] = `
<DocumentFragment>
  <header
    class="header"
    data-testid="header"
  >
    <div
      class="header__logo"
    >
      <a
        data-testid="logo"
        href="/"
      >
        <img
          class="logo"
        />
      </a>
    </div>
    <div
      class="header__about"
    >
      <div
        class="menu"
        data-testid="menu"
      >
        <ul
          class="menu__list"
        >
          <li
            class="menu__item"
          >
            <a
              href="/"
            >
              Home
            </a>
          </li>
          <li
            class="menu__item"
          >
            <a
              href=""
            >
              About
            </a>
          </li>
          <li
            class="menu__item"
          >
            <a
              href=""
            >
              Shop
            </a>
          </li>
          <li
            class="menu__item"
          >
            <a
              href=""
            >
              Blog
            </a>
          </li>
          <li
            class="menu__item"
          >
            <a
              href=""
            >
              FAQ
            </a>
          </li>
          <li
            class="menu__item"
          >
            <a
              href=""
            >
              Contact
            </a>
          </li>
        </ul>
      </div>
      <button
        class="btn btn__primary"
      >
        ORDER ONLINE
      </button>
    </div>
  </header>
  <figure>
    <img
      alt=""
      class="header__img"
    />
  </figure>
  <nav
    class="nav"
    data-testid="navigation"
  >
    <ul
      class="nav__list"
    >
      <li
        class="nav__item"
      >
        <a
          class="nav__link"
          href=""
        >
          <i
            class="fa fa-clock"
          />
           Mon - Sat: 8AM - 5PM
        </a>
      </li>
      <li
        class="nav__item"
      >
        <a
          class="nav__link"
          href=""
        >
          <i
            class="fa fa-phone"
          />
           123 456 7890
        </a>
      </li>
    </ul>
    <div
      class="search"
      data-testid="input-search"
    >
      <input
        class="search__text"
        placeholder="Search item"
        type="text"
      />
    </div>
  </nav>
  <div
    class="categories"
    data-testid="categories"
  >
    <p
      class="categories__title"
    >
      What are you looking for here?
    </p>
    <div
      class="categories"
      data-testid="category-list"
    >
      <ul
        class="categories__list"
      >
        <li
          class="categories__item active"
          data-testid="category-item"
        >
          All
        </li>
      </ul>
    </div>
  </div>
  <div
    class="viewProduct__list"
    data-testid="view-product-list"
  />
  <a
    class="viewProduct__link"
    href="/products"
  >
    <button
      class="btn btn__secondary"
    >
      VIEW ALL PRODUCTS
    </button>
  </a>
  <button
    class="btn__backToTop"
    data-testid="back-to-top"
  >
    <i
      class="fa fa-arrow-alt-circle-up"
    />
  </button>
  <div />
  <footer
    class="footer"
  >
    <p
      class="footer__contact"
      data-testid="footer"
    >
      Copyright © 2022 — Bakery WordPress Theme by Milingona.
    </p>
  </footer>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  />
</DocumentFragment>
`;
